### PR TITLE
FieldKind exposed as public API [API-437]

### PR DESCRIPTION
**Pull request by sancar**
Introduced FieldKind instead of FieldType which will be a
more correct API to show field id's of Compact and Portable.

FieldType has some unfortunate methods where typeSize needs to be
answered correctly. Since they differ between Portable and Compact,
we can not go with the same API.

The plan is to deprecate/remove FieldType together with Portable.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java [hunk 1]
```diff
@@ -222,7 +222,7 @@ public static SqlColumnMetadata createSqlColumnMetadata(String name, int type, b
     }
 
     public static FieldDescriptor createFieldDescriptor(@Nonnull String fieldName, int type) {
-        FieldType fieldType = FieldType.get((byte) type);
+        FieldKind fieldType = FieldKind.values()[type];
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java:225]: Minor: maybe change the variable name as well

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/FixedSizeTypesCodec.java [hunk 2]
```diff
@@ -82,8 +82,8 @@ public static void encodeInt(byte[] buffer, int pos, ProtocolType protocolType)
         encodeInt(buffer, pos, protocolType.getId());
     }
 
-    public static void encodeInt(byte[] buffer, int pos, FieldType fieldType) {
-        encodeInt(buffer, pos, fieldType.getId());
+    public static void encodeInt(byte[] buffer, int pos, FieldKind fieldKind) {
+        encodeInt(buffer, pos, fieldKind.ordinal());
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/FixedSizeTypesCodec.java:86]: I would advise not using `ordinal`, but introduce explicit ids for different field kinds. As Javadoc suggests, we should avoid using it. 
https://docs.oracle.com/javase/8/docs/api/java/lang/Enum.html#ordinal--

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java [hunk 3]
```diff
@@ -220,7 +220,7 @@ public ClientProxy getOrCreateProxy(@Nonnull String service,
         return getOrCreateProxyInternal(service, id, true);
     }
 
-    public ClientProxy getOrCreateLocalProxy(@Nonnull String service,
+    public ClientProxy  getOrCreateLocalProxy(@Nonnull String service,
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java:223]: Minor: unnecessary whitespace, let's revert this

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/GenericRecordQueryReader.java [hunk 4]
```diff
@@ -242,16 +242,16 @@ private Object readIndexed(InternalGenericRecord record, String path, int index)
         if (!record.hasField(path)) {
             return null;
         }
-        FieldType type = record.getFieldType(path);
-        return fieldOperations(type).readIndexed(record, path, index);
+        FieldKind id = record.getFieldKind(path);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/GenericRecordQueryReader.java:245]: Minor: We can rename this to `kind`. Same for below

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/AbstractGenericRecordBuilder.java [hunk 5]
```diff
@@ -39,193 +39,193 @@
...
+    protected abstract GenericRecordBuilder write(@Nonnull String fieldName, Object value, FieldKind fieldKind);
 
-    public static void checkTypeWithSchema(Schema schema, @Nonnull String fieldName, FieldType fieldType) {
+    public static void checkTypeWithSchema(Schema schema, @Nonnull String fieldName, FieldKind fieldKind) {
         FieldDescriptor fd = schema.getField(fieldName);
         if (fd == null) {
             throw new HazelcastSerializationException("Invalid field name: '" + fieldName + "' for " + schema);
         }
-        if (!fd.getType().equals(fieldType)) {
+        if (!fd.getKind().equals(fieldKind)) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/AbstractGenericRecordBuilder.java:226]: Minor: I guess it would be better to use `!=` on enums for equality checks

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/AbstractGenericRecordBuilder.java [hunk 6]
```diff
@@ -39,193 +39,193 @@
...
             throw new HazelcastSerializationException("Invalid field type: '" + fieldName
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/AbstractGenericRecordBuilder.java:227]: Let's fix the error message as well. It mentions `field type`

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactInternalGenericRecord.java [hunk 7]
```diff
@@ -187,12 +187,12 @@ public byte getByte(@Nonnull String fieldName) {
...
     }
 
-    boolean isFieldExists(@Nonnull String fieldName, @Nonnull FieldType type) {
+    boolean isFieldExists(@Nonnull String fieldName, @Nonnull FieldKind type) {
         FieldDescriptor field = schema.getField(fieldName);
         if (field == null) {
             return false;
         }
-        return field.getType().equals(type);
+        return field.getKind().equals(type);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactInternalGenericRecord.java:195]: Minor: Similar to above, I would use `==` and rename the variable from `type` to `kind`

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DefaultCompactWriter.java [hunk 8]
```diff
@@ -425,44 +425,44 @@ public void writeTimestampWithTimezoneArray(@Nonnull String fieldName, @Nullable
...
     }
 
-    protected FieldDescriptor checkFieldDefinition(@Nonnull String fieldName, FieldType fieldType) {
+    protected FieldDescriptor checkFieldDefinition(@Nonnull String fieldName, FieldKind fieldKind) {
         FieldDescriptor field = schema.getField(fieldName);
         if (field == null) {
             throw new HazelcastSerializationException("Invalid field name: '" + fieldName + " for " + schema);
         }
-        if (!field.getType().equals(fieldType)) {
+        if (!field.getKind().equals(fieldKind)) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DefaultCompactWriter.java:452]: Minor: `!=`

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DeserializedGenericRecord.java [hunk 9]
```diff
@@ -77,176 +77,176 @@ public boolean hasField(@Nonnull String fieldName) {
...
     }
 
-    private void check(@Nonnull String fieldName, FieldType fieldType) {
+    private void check(@Nonnull String fieldName, FieldKind fieldKind) {
         FieldDescriptor fd = schema.getField(fieldName);
         if (fd == null) {
             throw new HazelcastSerializationException("Invalid field name: '" + fieldName + " for " + schema);
         }
-        if (!fd.getType().equals(fieldType)) {
+        if (!fd.getKind().equals(fieldKind)) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/DeserializedGenericRecord.java:247]: Minor: `!=`

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/FieldDescriptor.java [hunk 10]
```diff
@@ -95,21 +95,21 @@ public boolean equals(Object o) {
...
+        result = 31 * result + fieldKind.hashCode();
         return result;
     }
 
     @Override
     public String toString() {
         return "FieldDefinitionImpl{"
                 + "fieldName='" + fieldName + '\''
-                + ", type=" + type
+                + ", type=" + fieldKind
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/FieldDescriptor.java:112]: `type` -> `kind` or `fieldKind`

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/RabinFingerprint.java [hunk 11]
```diff
@@ -51,7 +51,7 @@ public static long fingerprint64(Schema schema) {
         fingerPrint = fingerprint64(fingerPrint, schema.getFieldCount());
         for (FieldDescriptor descriptor : schema.getFields()) {
             fingerPrint = fingerprint64(fingerPrint, descriptor.getFieldName());
-            fingerPrint = fingerprint64(fingerPrint, descriptor.getType().getId());
+            fingerPrint = fingerprint64(fingerPrint, (byte) descriptor.getKind().ordinal());
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/RabinFingerprint.java:54]: Similar to comment above, let's avoid ordinal

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/Schema.java [hunk 12]
```diff
@@ -179,8 +179,8 @@ public void readData(ObjectDataInput in) throws IOException {
         for (int i = 0; i < fieldDefinitionsSize; i++) {
             String fieldName = in.readString();
             byte type = in.readByte();
-            FieldType fieldType = FieldType.get(type);
-            FieldDescriptor descriptor = new FieldDescriptor(fieldName, fieldType);
+            FieldKind fieldKind = FieldKind.values()[type];
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/Schema.java:182]: Using `values` is pretty inefficient as it allocates an array each time. We should introduce some logic on FieldKind to convert ids to FieldKinds

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java [hunk 13]
```diff
@@ -48,12 +49,12 @@ public boolean equals(Object o) {
             return false;
         }
         for (String fieldName : thatFieldNames) {
-            FieldType thatFieldType = that.getFieldType(fieldName);
-            FieldType thisFieldType = getFieldType(fieldName);
+            FieldKind thatFieldType = that.getFieldKind(fieldName);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:52]: Minor: Maybe rename variables

File: hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java [hunk 14]
```diff
@@ -187,7 +190,7 @@ private static boolean skipMethod(@Nonnull Class<?> clazz, @Nonnull Method metho
 
     @SuppressWarnings({"checkstyle:ReturnCount", "checkstyle:cyclomaticcomplexity"})
     @Nonnull
-    private static QueryDataType resolveType(@Nonnull FieldType type) {
+    private static QueryDataType resolveType(@Nonnull FieldKind type) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java:193]: Minor: rename `type`

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java (hunk 1 above)
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java:225]: Also, similar to comment below, we should cache the result of `values` and reuse it.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 15]
```diff
@@ -31,831 +31,829 @@
 import java.util.function.BiConsumer;
 
 /**
- * Should always be consistent with {@link com.hazelcast.nio.serialization.FieldType}
+ * Should always be consistent with {@link FieldKind}
  */
-@SuppressWarnings("checkstyle:AnonInnerLength")
-public enum FieldOperations {
+@SuppressWarnings({"checkstyle:AnonInnerLength", "checkstyle:ExecutableStatementCount"})
+public final class FieldOperations {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:37]: I personally find the old enum approach better for readability. Is there a reason we changed it to this?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 16]
```diff
@@ -884,4 +882,13 @@ public static FieldTypeBasedOperations fieldOperations(FieldType fieldType) {
...
     }
+
+    public static boolean isArrayType(FieldKind fieldKind) {
+        return fieldKind.ordinal() % 2 != 0;
+    }
+
+    public static FieldKind getSingleType(FieldKind fieldKind) {
+        assert isArrayType(fieldKind);
+        return FieldKind.values()[fieldKind.ordinal() - 1];
+    }
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:893]: :crossed_fingers: I hope this logic will hold in the future

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java (hunk 15 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:37]: I just though this is more straightforward. Not a strong reason.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 17]
```diff
@@ -170,9 +170,9 @@ private Object createObject(Class associatedClass) {
         return fields;
     }
 
-    private boolean fieldExists(Schema schema, String name, FieldType fieldType) {
+    private boolean fieldExists(Schema schema, String name, FieldKind fieldKind) {
         FieldDescriptor fieldDescriptor = schema.getField(name);
-        return fieldDescriptor != null && fieldDescriptor.getType().equals(fieldType);
+        return fieldDescriptor != null && fieldDescriptor.getKind().equals(fieldKind);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:175]:
```suggestion
        return fieldDescriptor != null && fieldDescriptor.getKind() == fieldKind;
```

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/Schema.java [hunk 18]
```diff
@@ -61,11 +61,11 @@ private void init() {
...
 
         for (FieldDescriptor descriptor : fieldDefinitionMap.values()) {
-            FieldType fieldType = descriptor.getType();
-            if (fieldOperations(fieldType).typeSizeInBytes() == VARIABLE_SIZE) {
+            FieldKind fieldKind = descriptor.getKind();
+            if (fieldOperations(fieldKind).kindSizeInBytes() == VARIABLE_SIZE) {
                 variableSizeFields.add(descriptor);
             } else {
-                if (FieldType.BOOLEAN.equals(fieldType)) {
+                if (FieldKind.BOOLEAN.equals(fieldKind)) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/Schema.java:68]:
```suggestion
                if (FieldKind.BOOLEAN == fieldKind) {
```

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/FieldTypeToFieldKind.java [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+                return FieldKind.TIMESTAMP;
+            case TIMESTAMP_ARRAY:
+                return FieldKind.TIMESTAMP_ARRAY;
+            case TIMESTAMP_WITH_TIMEZONE:
+                return FieldKind.TIMESTAMP_WITH_TIMEZONE;
+            case TIMESTAMP_WITH_TIMEZONE_ARRAY:
+                return FieldKind.TIMESTAMP_WITH_TIMEZONE_ARRAY;
+            default:
+                //This statement will never be called.
+                throw new IllegalStateException("There is no corresponding field id for given field type : " + fieldType);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/FieldTypeToFieldKind.java:98]:
```suggestion
                throw new IllegalStateException("There is no corresponding field kind for given field type : " + fieldType);
```

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java [hunk 20]
```diff
@@ -48,12 +49,12 @@ public boolean equals(Object o) {
             return false;
         }
         for (String fieldName : thatFieldNames) {
-            FieldType thatFieldType = that.getFieldType(fieldName);
-            FieldType thisFieldType = getFieldType(fieldName);
-            if (!thatFieldType.equals(thisFieldType)) {
+            FieldKind thatKind = that.getFieldKind(fieldName);
+            FieldKind thisKind = getFieldKind(fieldName);
+            if (!thatKind.equals(thisKind)) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:54]:
```suggestion
            if (thatKind != thisKind) {
```

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java [hunk 21]
```diff
@@ -70,14 +71,14 @@ public int hashCode() {
         int result = Objects.hash(getClassIdentifier());
         Set<String> thisFieldNames = getFieldNames();
         for (String fieldName : thisFieldNames) {
-            FieldType fieldType = getFieldType(fieldName);
+            FieldKind fieldType = getFieldKind(fieldName);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:74]: Leftover `fieldType`

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java [hunk 22]
```diff
@@ -70,14 +71,14 @@ public int hashCode() {
...
             result = 31 * result + fieldOperations(fieldType).hashCode(this, fieldName);
         }
         return result;
     }
 
     public final <T> T readAny(@Nonnull String fieldName) {
-        FieldType type = getFieldType(fieldName);
+        FieldKind type = getFieldKind(fieldName);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:81]: Leftover `type`

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java [hunk 23]
```diff
@@ -95,7 +96,7 @@ protected void writeFieldsToStringBuilder(StringBuilder stringBuilder) {
             i++;
             JsonEscape.writeEscaped(stringBuilder, fieldName);
             stringBuilder.append(": ");
-            FieldType fieldType = getFieldType(fieldName);
+            FieldKind fieldType = getFieldKind(fieldName);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:99]: Leftover `fieldType`

File: hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java [hunk 24]
```diff
@@ -177,7 +180,7 @@ private static boolean skipMethod(@Nonnull Class<?> clazz, @Nonnull Method metho
 
         // Add regular fields.
         for (String name : schema.getFieldNames()) {
-            FieldType compactType = schema.getField(name).getType();
+            FieldKind compactType = schema.getField(name).getKind();
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java:183]: Variable might be renamed

File: hazelcast/src/test/java/com/hazelcast/client/protocol/EnumCompatibilityTest.java [hunk 25]
```diff
@@ -132,42 +132,42 @@ public void testProtocolType() {
...
+        mappings.put(FieldKind.DATE_ARRAY, 23);
+        mappings.put(FieldKind.TIMESTAMP, 24);
+        mappings.put(FieldKind.TIMESTAMP_ARRAY, 25);
+        mappings.put(FieldKind.TIMESTAMP_WITH_TIMEZONE, 26);
+        mappings.put(FieldKind.TIMESTAMP_WITH_TIMEZONE_ARRAY, 27);
+        mappings.put(FieldKind.COMPACT, 28);
+        mappings.put(FieldKind.COMPACT_ARRAY, 29);
+        mappings.put(FieldKind.PORTABLE, 30);
+        mappings.put(FieldKind.PORTABLE_ARRAY, 31);
+        verifyCompatibility(FieldKind.values(), (Enum::ordinal), mappings);
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/protocol/EnumCompatibilityTest.java:170]: We could use `FieldKind::get` here

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java [hunk 26]
```diff
@@ -222,8 +222,8 @@ public static SqlColumnMetadata createSqlColumnMetadata(String name, int type, b
     }
 
     public static FieldDescriptor createFieldDescriptor(@Nonnull String fieldName, int type) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/CustomTypeFactory.java:224]: Minor: Maybe `type` -> `id`

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java [hunk 27]
```diff
@@ -140,8 +140,8 @@ void writeGenericRecord(BufferObjectDataOutput output, CompactGenericRecord reco
         Collection<FieldDescriptor> fields = schema.getFields();
         for (FieldDescriptor fieldDescriptor : fields) {
             String fieldName = fieldDescriptor.getFieldName();
-            FieldType fieldType = fieldDescriptor.getType();
-            fieldOperations(fieldType).writeFieldFromRecordToWriter(writer, record, fieldName);
+            FieldKind fieldKind = fieldDescriptor.getKind();
+            fieldOperations(fieldKind).writeFieldFromRecordToWriter(writer, record, fieldName);
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:144]: Question: `writeGenericRecord ` method reads from generic record and writes to writer buffer. The generic record keeps a buffer internally, is this buffer different from what will be the buffer for the writer? I was thinking that they are the same and generic record is just another way of accessing the fields. If they are the same, why not bulk copy the bytes?

Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:144]: Yes, this is one of the pending improvements that we can apply. There is more refactoring to be done when the buffer is from off-heap to on-heap or vice versa. I left those improvements to a later point to start simple.

**Review by ihsandemir — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 28]
```diff
@@ -31,831 +31,829 @@
...
 
-    FieldOperations(FieldTypeBasedOperations operations) {
-        this.operations = operations;
     }
 
-    public static FieldTypeBasedOperations fieldOperations(FieldType fieldType) {
-        FieldOperations fieldOperations = ALL[fieldType.getId()];
-        return fieldOperations.operations;
+    public static FieldKindBasedOperations fieldOperations(FieldKind fieldKind) {
+        return ALL[fieldKind.getId()];
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:856]: We are constructing the array with `.ordinal()` but getting with `.getId()`. I would prefer to use `.getId()` for both

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java (hunk 21 above)
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/nio/serialization/AbstractGenericRecord.java:74]: This still applies

**Review by mdumandag — APPROVED** [review]
Approving with minor comments

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java (hunk 28 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:856]: I prefer that as well. It was a leftover, I am changing it now.
